gateway/sds_gateway/api_methods/cog_pipelines.py

```diff
@@ -0,0 +1,654 @@
+        processed_data.save()
+
+    return processed_data
+
+
+def _process_waterfall_slice(params: WaterfallSliceParams) -> dict | None:
+    """Process a single waterfall slice."""
+    # Calculate sample range for this slice
+    slice_start_sample = (
+        params.start_sample + params.slice_idx * params.samples_per_slice
+    )
+    slice_num_samples = min(
+        params.samples_per_slice, params.end_sample - slice_start_sample
+    )
+
+    if slice_num_samples <= 0:
+        return None
+
+    # Read the data
+    data_array = params.reader.read_vector(
+        slice_start_sample, slice_num_samples, params.channel, 0
+    )
+
+    # Perform FFT processing
+    fft_data = np.fft.fft(data_array, n=params.fft_size)
+    power_spectrum = np.abs(fft_data) ** 2
+
+    # Convert to dB
+    power_spectrum_db = 10 * np.log10(power_spectrum + 1e-12)
+
+    # Convert power spectrum to binary string for transmission
+    data_bytes = power_spectrum_db.astype(np.float32).tobytes()
+    data_string = base64.b64encode(data_bytes).decode("utf-8")
+
+    # Create timestamp from sample index
+    timestamp = datetime.datetime.fromtimestamp(
+        slice_start_sample / params.sample_rate, tz=datetime.UTC
+    ).isoformat()
+
+    # Build WaterfallFile format with enhanced metadata
+    waterfall_file = {
+        "data": data_string,
+        "data_type": "float32",
+        "timestamp": timestamp,
+        "min_frequency": params.min_frequency,
+        "max_frequency": params.max_frequency,
+        "num_samples": slice_num_samples,
+        "sample_rate": params.sample_rate,
+        "center_frequency": params.center_freq,
+    }
+
+    # Build custom fields with all available metadata
+    custom_fields = {
+        "channel_name": params.channel,
+        "start_sample": slice_start_sample,
+        "num_samples": slice_num_samples,
+        "fft_size": params.fft_size,
+        "scan_time": slice_num_samples / params.sample_rate,
+        "slice_index": params.slice_idx,
+    }
+
+    waterfall_file["custom_fields"] = custom_fields
+    return waterfall_file
+
+
+def reconstruct_drf_files(capture, capture_files, temp_path: Path) -> Path | None:
+    """Reconstruct DigitalRF directory structure from SDS files."""
+    # Import utilities here to avoid Django app registry issues
+    from sds_gateway.api_methods.utils.minio_client import get_minio_client
+
+    logger.info("Reconstructing DigitalRF directory structure")
+
+    try:
+        minio_client = get_minio_client()
+
+        # Create the capture directory structure
+        capture_dir = temp_path / str(capture.uuid)
+        capture_dir.mkdir(parents=True, exist_ok=True)
+
+        # Download and place files in the correct structure
+        for file_obj in capture_files:
+            # Create the directory structure
+            file_path = capture_dir / file_obj.directory.lstrip("/") / file_obj.name
+            file_path.parent.mkdir(parents=True, exist_ok=True)
+
```

Comment:
as a safety measure, assert `file_path` (resolved) is a child of `temp_path` before creating the file

https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to

[example](https://github.com/spectrumx/sds-code/blob/0219977e9cfe88f95ae1adafc76f93610149ea55/gateway/sds_gateway/api_methods/helpers/reconstruct_file_tree.py#L286-L296)